components/tidb_query/src/codec/convert.rs

```diff
@@ -533,6 +537,137 @@ pub fn bytes_to_uint_without_context(bytes: &[u8]) -> Result<u64> {
     r.ok_or_else(|| Error::overflow("BIGINT UNSIGNED", ""))
 }
 
+pub fn produce_dec_with_specified_tp(
+    ctx: &mut EvalContext,
+    mut dec: Decimal,
+    ft: &FieldType,
+) -> Result<Decimal> {
+    let (flen, decimal) = (ft.flen(), ft.decimal());
+    if flen != UNSPECIFIED_LENGTH && decimal != UNSPECIFIED_LENGTH {
+        if flen < decimal {
+            return Err(Error::m_bigger_than_d(""));
+        }
+        let (prec, frac) = dec.prec_and_frac();
+        let (prec, frac) = (prec as isize, frac as isize);
+        if !dec.is_zero() && prec - frac > flen - decimal {
+            // select (cast 111 as decimal(1)) causes a warning in MySQL.
+            ctx.handle_overflow(Error::overflow(
+                "Decimal",
+                &format!("({}, {})", flen, decimal),
+            ))?;
+            dec = max_or_min_dec(dec.is_negative(), flen as u8, decimal as u8)
+        } else if frac as isize != decimal {
+            let old = dec.clone();
+            let rounded = match dec.round(decimal as i8, RoundMode::HalfEven) {
+                Res::Ok(d) => d,
+                Res::Truncated(_) => {
+                    return Err(Error::truncated());
+                }
+                Res::Overflow(_) => {
+                    // TODO, is the err msg right
+                    return Err(Error::overflow(
+                        "Decimal",
+                        &format!("({}, {})", flen, decimal),
+                    ));
+                }
+            };
+            if !rounded.is_zero() && frac > decimal && rounded != old {
+                if ctx.cfg.flag.contains(Flag::IN_INSERT_STMT)
+                    || ctx.cfg.flag.contains(Flag::IN_UPDATE_OR_DELETE_STMT)
+                {
+                    ctx.warnings.append_warning(Error::truncated());
+                } else {
+                    if let Err(e) = ctx.handle_truncate(true) {
+                        if e.is_overflow() {
+                            ctx.handle_overflow(e)?
+                        }
+                    }
+                }
+            }
+            dec = rounded
+        }
+    };
+    if ft.is_unsigned() && dec.is_negative() {
+        Ok(Decimal::from(0 as u64))
```

Comment:
Prefer to use `Decimal::zero()`.

Reply:
It seems there is a difference from them.

Reply:
`assert_eq!(Decimal::from(0_u64), Decimal::zero());` will pass, you can try it.